File: src/main/java/cookiedragon/luchadora/integration/discord/DiscordIntegration.java
package cookiedragon.luchadora.integration.discord;

import com.jagrosh.discordipc.IPCClient;
import com.jagrosh.discordipc.IPCListener;
import com.jagrosh.discordipc.entities.Packet;
import com.jagrosh.discordipc.entities.User;
import cookiedragon.luchadora.integration.IIntegration;
import org.json.JSONObject;

/**
 * @author cookiedragon234 22/Dec/2019
 */
public class DiscordIntegration implements IIntegration
{
	IPCClient ipcClient;
	
	@Override
	public void start()
	{
		ipcClient = new IPCClient(658340490655039537L);
		ipcClient.setListener(new IPCListener()
		{
			@Override
			public void onReady(IPCClient client)
			{
				System.out.println("Connected to discord " + client.getDiscordBuild());
				
				DiscordPrecenceThread.start(client);
				System.out.println("Started Discord RPC");
			}
			
			@Override
			public void onActivityJoin(IPCClient client, String secret)
			{
				System.out.println("Requesting to join activity " + secret);
			}
			
			@Override
			public void onClose(IPCClient client, JSONObject json)
			{
				System.out.println("Discord connection closed");
			}
			
			@Override
			public void onDisconnect(IPCClient client, Throwable t)
			{
				System.out.println("Discord connection disconnected");
			}
		});
		try
		{
			ipcClient.connect();
		}
		catch (Exception e)
		{
			new RuntimeException("Failure Starting RP", e).printStackTrace();
		}
	}
}

File: src/main/kotlin/cookiedragon/luchadora/integrations/discord/DiscordIntegration.kt
package cookiedragon.luchadora.integrations.discord

import com.jagrosh.discordipc.IPCClient
import com.jagrosh.discordipc.IPCListener
import com.jagrosh.discordipc.entities.RichPresence
import cookiedragon.luchadora.managers.ForgeEventListener.lastChat
import cookiedragon.luchadora.util.Globals
import cookiedragon.luchadora.util.Globals.mc
import cookiedragon.luchadora.util.Initialisable
import org.json.JSONObject
import kotlin.concurrent.thread

/**
 * @author cookiedragon234 17/Feb/2020
 */

object DiscordIntegration : Initialisable() {
	private val ipcClient: IPCClient = IPCClient(658340490655039537L)
	private var richPresence: RichPresence? = null
	
	init {
		ipcClient.setListener(object : IPCListener {
			override fun onReady(client: IPCClient) {
				println("Connected to discord " + client.discordBuild)
				while (true) {
					try {
						richPresence = RichPresence.Builder()
							.setState(getState())
							.setDetails(getDetails())
							.setLargeImage("luchadora", "Luchadora")
							.build()
					} catch(e: Exception) {
						e.printStackTrace()
					} finally {
						if (richPresence != null) client.sendRichPresence(richPresence)
					}
					try { Thread.sleep(3500) } catch (ignored: Exception) {}
				}
			}
			
			override fun onClose(client: IPCClient, json: JSONObject) {
				println("Discord connection closed")
			}
			
			override fun onDisconnect(client: IPCClient, t: Throwable) {
				println("Discord connection disconnected")
			}
		})
		try {
			thread(true, isDaemon = true) {
				ipcClient.connect()
			}
			println("Started Discord RPC")
		} catch (e: Exception) {
			e.printStackTrace()
		}
	}
	
	private val discord = "discord/fUkXhEu"
	fun getState(): String {
		val ip = mc.currentServerData?.serverIP ?: ""
		return when {
			mc.world == null -> discord
			mc.isSingleplayer -> "SinglePlayer"
			ip.contains("2b2t") -> {
				if (lastChat?.startsWith("Position in queue: ") == true)
					"$ip ${lastChat?.substring("Position in queue: ".length)} in q"
				else ip
			}
			else -> ip
		}
	}
	
	fun getDetails(): String {
		return when {
			mc.isSingleplayer -> "SinglePlayer"
			mc.world == null -> "Main Menu"
			else -> "Multiplayer"
		}
	}
}

